pyecsca/sca/target/emulator.py

```diff
@@ -0,0 +1,123 @@
+        temp_trace = []
+        context.actions.walk(callback)
+        return Trace(np.array(temp_trace))
+    
+    def emulate_scalar_mult_traces(self, num_of_traces: int, scalar: int) -> Tuple[list[Point], list[Trace]]:
+        points = [self.params.curve.affine_random().to_model(self.coords, self.params.curve) for _ in range(num_of_traces)]
+        traces = []
+        for point in points:
+            _, trace = self.scalar_mult(scalar, point)
+            traces.append(trace)
+        return points, traces
+
+    def emulate_ecdh_traces(self, num_of_traces: int) -> Tuple[list[Trace], list[Trace]]:
+        other_pubs = [self.params.curve.affine_random().to_model(self.coords, self.params.curve) for _ in range(num_of_traces)]
+        traces = []
+        for pub in other_pubs:
+            _, trace = self.ecdh(pub, None)
+            traces.append(trace)
+        return other_pubs, traces
+    
+    def set_params(self, params: DomainParameters) -> None:
+        self.params = params
+
+    def set_leakage_model(self, leakage_model: LeakageModel) -> None:
+        self.leakage_model = leakage_model
+
+    def scalar_mult(self, scalar: int, point: Point) -> Tuple[Point, Trace]:
+        with local(DefaultContext()) as ctx:
+            self.mult.init(self.params, point)
+            res_point = self.mult.multiply(scalar)
+        return res_point, self.get_trace(ctx)
+
+    def generate(self) -> Tuple[Tuple[Mod, Point], Trace]:
+        with local(DefaultContext()) as ctx:
+            keygen = KeyGeneration(self.mult, self.params, False)
+            priv, pub =  keygen.generate()
+        return (priv, pub), self.get_trace(ctx)
+
+    def set_privkey(self, privkey: int) -> None:
+        self.privkey = privkey
+
+    def set_pubkey(self, pubkey: Point) -> None:
+        self.pubkey = pubkey
+
+    def ecdh(self, other_pubkey: Point, hash_algo=None) -> Tuple[bytes, Trace]:
+        with local(DefaultContext()) as ctx:
+            ecdh = KeyAgreement(self.mult, self.params, other_pubkey, self.privkey, hash_algo)
+            shared_secret = ecdh.perform()
+        return shared_secret, self.get_trace(ctx)
+
+    def ecdsa_sign(self, data: bytes, hash_algo=None) -> Tuple[bytes, Trace]:
+        with local(DefaultContext()) as ctx:
+            ecdsa = Signature(self.mult, self.params, self.mult.formulas["add"], self.pubkey, self.privkey, hash_algo)
+            signed_data = ecdsa.sign_data(data)
+        return signed_data, self.get_trace(ctx)
+
+    def ecdsa_verify(self, data: bytes, signature: bytes, hash_algo=None) -> Tuple[bool, Trace]:
+        with local(DefaultContext()) as ctx:
+            ecdsa = Signature(self.mult, self.params, self.mult.formulas["add"], self.pubkey, hash_algo) 
+            verified = ecdsa.verify_data(signature, data)
+        return verified, self.get_trace(ctx)
+
+    def debug(self):
+        return self.model.shortname, self.coords.name
+
+    def connect(self):
+        super().connect()
```

Comment:
Don't do this. Just pass.
```suggestion
        pass
```